### PR TITLE
Change the 'Use auto property' code fixer to not be a 'compilation end' analyzer.

### LINKED ISSUE #20855: IDE0032 Use Auto Property code fix mangles code in preexisting property with getter
**Version Used**: d15rel 26711.00

**Steps to Reproduce**:

```
git clone https://github.com/microsoft/vs-threading.git
cd vs-threading
git checkout 0404d41776090ad54fd892081dcd1bd2b9165698
cd src
MSBuild /t:restore
.\Microsoft.VisualStudio.Threading.sln
```

1. Once in VS with the solution open, navigate to AsyncQueue.cs.
1. (I had netstandard1.1 selected as the language service context)
1. When the language service fades out the first field in the class (`syncObject`) activate the "Use Auto Property" code fix.

**Expected Behavior**:

No compile errors and no incorrect changes to code.

**Actual Behavior**:

The `SyncRoot` property which exposed the `syncObject` property is badly mangled. It was this:

```csharp
        protected object SyncRoot
        {
            get { return this.syncObject; }
        }
```

And was changed to this (including the comments, invalid syntax, etc.):

```csharp
        protected object SyncRoot {
/* Unmerged change from project 'Microsoft.VisualStudio.Threading(netstandard1.1)'
Before:
            get { return this.syncObject; }
After:
            get { return set; }
*/
 get; private this.SyncRoot; }
```

Note that there are absolutely no `#if` directives anywhere in the file. But this file is in a project that has 3 target frameworks.

### LINKED ISSUE #23215: IDE0032 (Use Auto Property) does not work when full solution analysis is disabled
**Version Used**: 15.5 Preview 3

🔗 Observed while investigating [this community report](https://developercommunity.visualstudio.com/content/problem/150002/ide0032-performs-less-than-ideal.html).

### Steps to Reproduce

1. Add a property:

    ```csharp
    private string _text;

    public string Text
    {
      get { return _text; }
      set { _text = value; }
    }
    ```

### Expected Behavior

The `_text` field is grayed (unnecessary code), and a code fix is provided to convert the field to an auto property.

### Actual Behavior

No suggestion is given.

### Additional details

If **Full Solution Analysis** is enabled, the feature works as expected. However, this feature is disabled both by default and by recommendation, with no plans to enable it. IDE analyzers cannot rely on this feature being enabled.

### LINKED ISSUE #23216: IDE0032 (Use Auto Property) Fix All does not work
**Version Used**: 15.5 Preview 3

🔗 Observed while investigating [this community report](https://developercommunity.visualstudio.com/content/problem/150002/ide0032-performs-less-than-ideal.html).

### Steps to Reproduce

1. Attempt to apply a Fix All operation to IDE0032 (Use Auto Property)

### Expected Behavior

Fix All, when available, fixes all instances of an issue within the specified context.

### Actual Behavior

Only one instance is fixed.

### Additional Details

* This behavior is observed even when Full Solution Analysis is enabled (see #23215)
* This behavior is observed even when IDE0032 is set to Warning severity

### LINKED ISSUE #23217: Disable IDE0032 until it works
**Version Used**: 15.5 Preview 3

🔗 Observed while investigating [this community report](https://developercommunity.visualstudio.com/content/problem/150002/ide0032-performs-less-than-ideal.html).

**Steps to Reproduce**:

1. Attempt to use the Use Auto Property diagnostic and code fix

**Expected Behavior**:

1. Code does not break
2. The code fix works
3. The fix all operation either work or is not provided

**Actual Behavior**:

I was unable to use this feature in any context and have it provide a meaningful result.

* The feature was not even visible without enabling a non-default (and not recommended) feature (#23215, #17857)
* The code was broken when the feature was used (#20855)
* The code fix didn't perform as expected (#23216)

This diagnostic needs to be disabled until it meets a reasonable quality bar.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes #23216
Fixes #23215
Fixes #20855
Closes #23217

This PR changes the 'use auto property' code fixer to work as a Semantic-Model-Analyzer, instead of a Compilation-End analyzer.  Compilation-End-Analyzers are basically an unusable part of the analyzer API for code that wants to run in VS.  They are too expensive, and users can't interact with their results in a timely manner, like they can with normal analyzers.

As part of making this happen, I had to make a small change to the functionality of the feature.  Specifically, we used to determine if it was safe to make the change by analyzing all usages of a the field-to-remove in the project the feature was analyzing.  This was very expensive, but was necessary in case some code was using the field in an unexpected manner.  Now, we limit the feature to only private fields.  This means we can do that analysis in a timely manner, as we only have to examine the type it is contained in.  

This should have very little end user impact as the majority of user fields are private (esp. the ones that are paired with a property that you'd want ot make an auto prop out of).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell @Pilchie @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Note: this doesn't fix https://github.com/dotnet/roslyn/issues/20855

It's not clear to me what's going on with that code.  There seems to be some very strange code related to merging in changes with linked files.  There are linked files in the reported case.  But i have no idea what this linked-file-merging code is trying to do.  @dpoeschl Do you have any context to shed some light here?

**Comment by CyrusNajmabadi** [comment]
I've investigated #20855 a little further.  The linked file merging step is borking because it sees three files being changed, but sees a different set of edits applied to each file (wat).  

For example, given AsyncQueue.cs in project1, it sees both the removal of the field, the change of the property, and the rename of references to the field to now point at the property.  However, for another linked AsyncQueue.cs in project2, it only sees the updates of the field references to point at the property.  However, the field and property still remain.  

My guess is that this has to do with how UseAutoProp updates the solution.  Investigating further.

**Comment by CyrusNajmabadi** [comment]
Ok.  I figure it out.  FindRefs+Rename+LinkedDocs = Wacky.

**Comment by CyrusNajmabadi** [comment]
Test added for the linked files case.

File: src/EditorFeatures/Test2/Diagnostics/UseAutoProperty/UseAutoPropertyTests.vb [hunk 1]
```diff
@@ -92,5 +94,54 @@ end class
 </text>.Value.Trim()}
                 })
         End Function
+
+        <Fact(), Trait(Traits.Feature, Traits.Features.CodeActionsUseAutoProperty)>
+        Public Async Function TestLinkedFile() As System.Threading.Tasks.Task
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/Diagnostics/UseAutoProperty/UseAutoPropertyTests.vb]: Validated this test fails before my change (showing the 'linked file marge' comment that @AArnott saw).

File: src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs [hunk 2]
```diff
@@ -12,48 +14,100 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty
...
             return variable.Initializer?.Value;
         }
 
-        private void AnalyzeArgument(ConcurrentBag<IFieldSymbol> ineligibleFields, SyntaxNodeAnalysisContext context)
+        private void AnalyzeArgument(
+            SemanticModel semanticModel, ArgumentSyntax argument,
+            HashSet<IFieldSymbol> ineligibleFields, CancellationToken cancellationToken)
         {
             // An argument will disqualify a field if that field is used in a ref/out position.  
             // We can't change such field references to be property references in C#.
```
Comment (sharwell) [inline src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: ❗️ Need to handle case where a struct is passed to a `ref this` extension method (either as part of this pull request, or by creating a new issue and replying to this comment with the new issue number).

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: Done: https://github.com/dotnet/roslyn/issues/23257

File: src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs [hunk 3]
```diff
@@ -12,48 +14,100 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UseAutoProperty
...
+            }
+
+            // If we have a class or struct, recurse inwards.
+            if (member.IsKind(SyntaxKind.ClassDeclaration, out TypeDeclarationSyntax typeDeclaration) ||
+                member.IsKind(SyntaxKind.StructDeclaration, out typeDeclaration))
+            {
+                AnalyzeMembers(context, typeDeclaration.Members, analysisResults);
+            }
+
+            if (member is PropertyDeclarationSyntax propertyDeclaration)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: Good call.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: #done.

[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/CSharp/UseAutoProperty/UseAutoPropertyAnalyzer.cs]: ❔ Were you going to make the update or want me to?

**Comment by sharwell** [comment]
@MattGertz for ask mode

**Pull request merged**